### PR TITLE
Don't launch ForAll kernels with 0 elements (Fixes #5017)

### LINKED ISSUE #5017: `cuda.to_device()` cannot copy zero-element data
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Not sure if this is worth fixing. I found it when trying to use Numba CUDA arrays as the backend for mpi4py's GPU tests. Most likely to fix this we need a check in `to_device()` before copying?
```python
>>> from numba import cuda
>>> # in our tests we loop over N from 0 to a fixed positive integer,
... # and only N=0 caused the error
>>> N = 0  
>>> a = range(N)  
>>> arr = cuda.device_array(len(a), 'f')
>>> arr[:] = cuda.to_device(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/devices.py", line 225, in _require_cuda_context
    return fn(*args, **kws)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/devicearray.py", line 530, in __setitem__
    return self._do_setitem(key, value)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/devicearray.py", line 581, in _do_setitem
    _assign_kernel(lhs.ndim).forall(n_elements, stream=stream)(lhs, rhs)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/compiler.py", line 275, in __call__
    sharedmem=self.sharedmem)(*args)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/compiler.py", line 537, in __call__
    sharedmem=self.sharedmem)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/compiler.py", line 611, in _kernel_call
    cu_func(*kernelargs)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/driver.py", line 1578, in __call__
    self.sharedmem, streamhandle, args)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/driver.py", line 1622, in launch_kernel
    None)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [1] Call to cuLaunchKernel results in CUDA_ERROR_INVALID_VALUE
```

### PR DESCRIPTION
**Pull request by gmarkall**
When created with 0 elements, a `ForAll` would configure a kernel with 0 blocks and then cause a `CudaAPIError` when attempting to launch. This commit fixes this issue by returning early from `ForAll.__call__` when the `ForAll` was constructed with 0 elements.

Tests are added for the original issue reported, where assigning a zero-length slice would trigger the error, and also for directly constructing a `ForAll` with 0 elements.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. Seems like there's a test fail: https://dev.azure.com/numba/numba/_build/results?buildId=3769&view=logs&j=eb149424-352d-52d9-b16d-aeffaad6858a&t=cdac5993-002e-5764-1ca7-e646ebe78687&l=244 that could do with looking at. Thanks!

**Comment by gmarkall** [comment]
It is now fixed :-)

**Comment by seibert** [comment]
internal CI: `numba_smoketest_cuda_7`

**Comment by seibert** [comment]
Passed internal CI

File: numba/cuda/simulator/cudadrv/devicearray.py [hunk 1]
```diff
@@ -188,6 +188,7 @@ def check_array_compatibility(ary1, ary2):
 
 
 def to_device(ary, stream=0, copy=True, to=None):
+    ary = np.array(ary, copy=False, subok=True)
```
Comment (stuartarchibald) [inline numba/cuda/simulator/cudadrv/devicearray.py]: note to self. replicates https://github.com/numba/numba/blob/9a3f13020475ea945c7f0359db3e3350484710d6/numba/cuda/cudadrv/devicearray.py#L699 semantics

File: numba/cuda/compiler.py [hunk 2]
```diff
@@ -261,6 +261,9 @@ def __init__(self, kernel, ntasks, tpb, stream, sharedmem):
         self.sharedmem = sharedmem
 
     def __call__(self, *args):
+        if self.ntasks == 0:
+            return
```
Comment (stuartarchibald) [inline numba/cuda/compiler.py]: Should this also handle `.ntasks < 0` and raise, else it'll just end up as an error at kernel launch?

Comment (gmarkall) [reply numba/cuda/compiler.py]: Yes, it probably should. Will add this.

Comment (gmarkall) [reply numba/cuda/compiler.py]: I've decided to guard for this in the initialiser instead - it seems to me that it's legitimate for a `ForAll` to have no work to do and therefor the call to it be a no-op, but it's not legitimate to even create a `ForAll` with negative work.

**Comment by seibert** [comment]
Looks like this fails to raise in the simulator case?

**Comment by seibert** [comment]
looks like everything is passing CI now and ready for re-review

**Pull request merged**